File: rusty_web/src/api/projects.rs
use serde_json::Value;

use commons::errors::RustyError;
use domain::projects::{PagedProjects, Project, RegisterProject};

use crate::api::client::reqwasm_post;
use crate::api::utils::{parse_entries, parse_paged};

/// Function to retrieve projects from a GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_projects() -> Result<PagedProjects, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            projects {{
                get(options: {{ pageSize: 99 }}) {{
                    total
                    page
                    pageSize
                    entries {{
                        id
                        name
                        url
                    }}
                }}
            }}
        }}"#),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["projects"]["get"].clone();
    let (total, page, page_size, entries) = parse_paged(&json_data)?;
    let entries = parse_entries(entries)?;
    Ok(PagedProjects {
        total,
        page,
        page_size,
        entries,
    })
}

/// Function to retrieve a project from a GraphQL endpoint by id.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_project(id: String) -> Result<Project, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            projects {{
                getById(id: "{}") {{
                    id
                    name
                    url
                }}
            }}
        }}"#, id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["projects"]["getById"].clone();
    parse_entries(json_data)
}

/// Function to register a new project via GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn register_project(model: RegisterProject) -> Result<String, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"mutation {{
            projects {{
                register(project: {{
                    name: "{}",
                    url: "{}"
                }})
            }}
        }}"#, model.name, model.url),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: Value = serde_json::from_str(&data)?;
    Ok(json_data["data"]["projects"]["register"].to_string())
}

File: rusty_web/src/lib/api/projects.ts
import { fetchPost } from '$lib/utils/api';
import { type Project, projectSourceMap } from '$lib/domain/project';

const getProjectsQuery = (page: number, group: string, source: string, name: string) => {
	const groupMatch = group.match(/[a-z-]/gi);
	const groupIdFilter = `group_id: ${groupMatch === null ? null : `{ equals: ${group} }`}`;
	const nameFilter = `name: { contains: "${name === undefined ? '' : name}" }`;
	const filter = `filter: { ${groupIdFilter}, ${nameFilter}, source: { equals: "${projectSourceMap[source]}" } }, `;
	const options = `options: { pageNumber: ${page}, pageSize: 30, sortMode: ASCENDING, sortField: "name" }`;

	return `query {
		projects {
			get(${filter}${options}){
				total
				page
				pageSize
				entries {
					id
					source
					name
					url
					jobs {
						name
						pipelines {
							id
							number
							status
							stageStatus
							registerDate
						}
					}
				}
			}
		}
	}`;
};

export const fetchProjects = async (
	auth: string,
	page: number,
	group: string,
	source: string,
	name: string
) => {
	try {
		const response = await fetchPost(
			auth,
			JSON.stringify({ query: getProjectsQuery(page, group, source, name) })
		);

		if (!response.ok) {
			return {
				errors: ['Fetch projects failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				const paged = data?.projects?.get;
				const projects: Project[] = paged?.entries ?? [];
				projects.forEach((p) => {
					const flattened = p.jobs.reduce((accumulator: any[], current: any) => {
						current.pipelines.forEach((pp: any) => (pp.jobName = current.name));
						return accumulator.concat(current.pipelines);
					}, []);
					flattened.sort((a, b) => {
						if (a.registerDate < b.registerDate) {
							return 1;
						}
						if (a.registerDate > b.registerDate) {
							return -1;
						}
						return 0;
					});
					if (flattened.length > 0) {
						p.lastPipeline = flattened[0];
					}
				});

				return {
					total: paged?.total ?? 0,
					page: paged?.page ?? 1,
					pageSize: paged?.pageSize ?? 20,
					entries: projects
				};
			}
		}
	} catch (error) {
		return {
			errors: ['Fetch projects failed']
		};
	}
};

const getProjectByIdQuery = (id: string) => {
	return `query {
		projects {
			getById(id: "${id}"){
				id
				source
				name
				url
			}
		}
	}`;
};

export const getProjectById = async (auth: string, id: string) => {
	try {
		const response = await fetchPost(auth, JSON.stringify({ query: getProjectByIdQuery(id) }));

		if (!response.ok) {
			return {
				errors: ['Get project by id failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				return data?.projects?.getById;
			}
		}
	} catch (error) {
		return {
			errors: ['Get project by id failed']
		};
	}
};
